### PR TITLE
Use queues for the pod events

### PR DESCRIPTION
**Pull request by ikitiki**
Make use of the FIFO queue for the pod events of the cluster

### CODE REVIEW

File: pkg/cluster/cluster.go [hunk 1]
```diff
@@ -460,7 +463,38 @@ func (c *Cluster) Delete() error {
 }
 
 func (c *Cluster) ReceivePodEvent(event spec.PodEvent) {
-	c.podEvents <- event
+	c.podEventsQueue.Add(event)
+}
+
+func (c *Cluster) podEventProcess(obj interface{}) error {
```
Comment (alexeyklyukin) [inline pkg/cluster/cluster.go]: I'd rename it to processPodEvent (action -> object, i.e. setStatus), or name it podEventProcessor.

File: pkg/cluster/cluster.go [hunk 2]
```diff
@@ -460,7 +463,38 @@ func (c *Cluster) Delete() error {
...
+	}
+
+	return nil
+}
+
+func (c *Cluster) Run(stopCh <-chan struct{}) {
+	go c.processPodEventQueue(stopCh)
+}
+
+func (c *Cluster) processPodEventQueue(stopCh <-chan struct{}) {
```
Comment (alexeyklyukin) [inline pkg/cluster/cluster.go]: Can we drop it and embed its body into the Run?

File: pkg/cluster/cluster.go [hunk 3]
```diff
@@ -460,7 +463,38 @@ func (c *Cluster) Delete() error {
...
+	c.podSubscribersMu.RUnlock()
+	if ok {
+		subscriber <- event
+	}
+
+	return nil
+}
+
+func (c *Cluster) Run(stopCh <-chan struct{}) {
+	go c.processPodEventQueue(stopCh)
```
Comment (alexeyklyukin) [inline pkg/cluster/cluster.go]: I'd make it, perhaps, return the *Cluster, so that one can write cl := cluster.New(....).Run()

**Comment by alexeyklyukin** [comment]
Minor stylistic comments apart LGTM.

**Comment by alexeyklyukin** [comment]
👍

File: pkg/cluster/cluster.go (hunk 2 above)
Comment (ikitiki) [inline pkg/cluster/cluster.go]: I want to have a place where I can run all the future go routines.
besides, lambda func will not look cool in the Run() function

Comment (alexeyklyukin) [inline pkg/cluster/cluster.go]: it there will be more - let's leave it as it is now.

**Comment by alexeyklyukin** [comment]
👍

**Comment by ikitiki** [comment]
+1

**Comment by ikitiki** [comment]
👍

File: pkg/cluster/cluster.go [hunk 4]
```diff
@@ -460,7 +463,38 @@ func (c *Cluster) Delete() error {
...
+
+func (c *Cluster) podEventProcess(obj interface{}) error {
+	event, ok := obj.(spec.PodEvent)
+	if !ok {
+		return fmt.Errorf("could not cast to PodEvent")
+	}
+
+	c.podSubscribersMu.RLock()
+	subscriber, ok := c.podSubscribers[event.PodName]
+	c.podSubscribersMu.RUnlock()
```
Comment (ikitiki) [inline pkg/cluster/cluster.go]: yeah. good point

Comment (alexeyklyukin) [inline pkg/cluster/cluster.go]: @Jan-M I'd not use defer here. One should hold the lock for as short as possible, and avoid holding it over the statements and does the waiting on their own, like reading from the channel.  Besides, there is nothing to fail in the statement between the lock and unlock.

Comment (ikitiki) [inline pkg/cluster/cluster.go]: JFYI, https://github.com/beyondns/gotips/blob/master/tips32.md#23---defer-is-slow

**Pull request merged**